src/JBGovernanceNFT.sol

```diff
@@ -1,159 +1,120 @@
         address _sender = _msgSender();
         for (uint256 _i; _i < _mints.length;) {
-            // Should never be more than a uint200 or 0
-            if (_mints[_i].stakeAmount == 0 || _mints[_i].stakeAmount > type(uint200).max) {
+            // Should never be 0
+            if (_mints[_i].stakeAmount == 0) {
                 revert INVALID_STAKE_AMOUNT(_i, _mints[_i].stakeAmount);
             }
+
+            stakingTokenBalance[_mints[_i].beneficiary] += _mints[_i].stakeAmount;
+
+            // Living on the edge, using safemint because we can
+            _safeMint(_mints[_i].beneficiary, nextokenId);
+    
+             _tokenId = nextokenId;
+
+            emit NFT_Minted_and_Staked(nextokenId, _sender);
+
             // Transfer the stake amount from the user
             token.safeTransferFrom(_sender, address(this), _mints[_i].stakeAmount);
+
             // Get the tokenId to use and increment it for the next usage
-            unchecked {
-                _tokenId = ++nextokenId;
-            }
-            // If the NFT should staked immediatly then we do so
-            // otherwise stakedAt becomes address(0) which means unstaked
-            address _stakedAt;
-            if (_mints[_i].stakeNFT) {
-                _stakedAt = _sender;
-                emit NFTStaked(_tokenId, _sender);
-            }
-            // Store the info regarding this staked position
-            stakes[_tokenId] = JBGovernanceNFTStake({
-                amount: uint200(_mints[_i].stakeAmount),
-                stakedAt: _stakedAt
-            });
-            // Living on the edge, using safemint because we can
-            _safeMint(_mints[_i].beneficiary, _tokenId);
             unchecked {
                 ++_i;
+                nextokenId += 1;
             }
         }
     }
 
-    function burn(JBGovernanceNFTBurn[] calldata _burns) external {
+    /**
+     * @dev Burn the nft and unstake tokens.
+     * @param _burns An array of struct containing the token id and beneficiary to be sent the staked tokens too.
+    */
+    function burn_and_unstake(JBGovernanceNFTBurn[] calldata _burns) external {
         for (uint256 _i; _i < _burns.length;) {
             uint256 _tokenId = _burns[_i].tokenId;
+            address _owner = _ownerOf(_tokenId);
+            address _beneficiary = _burns[_i].beneficiary;
+
             // Make sure only the owner can do this
-            if (_ownerOf(_tokenId) != msg.sender) 
-                revert NO_PERMISSION(_tokenId);
-            // If the nft is staked it has to be unstaked first
-            if(stakes[_tokenId].stakedAt != address(0))
-                revert NFT_IS_STAKED(_tokenId, stakes[_tokenId].stakedAt);
-            // Immedialty burn to prevernt reentrency
+            if (_owner != msg.sender) 
+                revert NO_PERMISSION(_tokenId);  
+            
+            uint256 _stakeAmount = stakingTokenBalance[_beneficiary];
```

Comment:
We should track the token amount for each token, if a user has 2 tokens and they burn 1 of them this will send them the entire staked balance.

Reply:
Additionally this also unstakes the balance of `_beneficiary` and not the one from `_owner`

Reply:
regarding unstaking balance of `_beneficiary` is because since `_beneficiary` is owner of the nft hence while staking and since the nft is only burnable by the owner